File: geomesa-index-api/src/main/java/org/locationtech/geomesa/index/conf/TableSplitter.java
/***********************************************************************
 * Copyright (c) 2013-2023 Commonwealth Computer Research, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.index.conf;

import org.opengis.feature.simple.SimpleFeatureType;

public interface TableSplitter {

    /**
     * Get splits for a table
     *
     * @param sft simple feature type
     * @param index name of the index being configured
     * @param options splitter options
     * @return split points
     */
    byte[][] getSplits(SimpleFeatureType sft, String index, String options);

    /**
     * Get splits for a partitioned table
     *
     * @param sft simple feature type
     * @param index name of the index being configured
     * @param partition name of the partition being configured
     * @param options splitter options
     * @return split points
     */
    byte[][] getSplits(SimpleFeatureType sft, String index, String partition, String options);
}

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/conf/splitter/TableSplitter.scala
/***********************************************************************
 * Copyright (c) 2013-2020 Commonwealth Computer Research, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.index.conf.splitter

import org.locationtech.geomesa.index.conf.TableSplitter
import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes.Configs.{TableSplitterClass, TableSplitterOpts}
import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes.InternalConfigs.{PartitionSplitterClass, PartitionSplitterOpts}
import org.opengis.feature.simple.SimpleFeatureType

/**
  * Companion object for TableSplitter java interface
  */
object TableSplitter {

  def getSplits(sft: SimpleFeatureType, index: String, partition: Option[String] = None): Array[Array[Byte]] = {
    partition match {
      case None =>
        val splitter = create(sft.getUserData.get(TableSplitterClass).asInstanceOf[String])
        splitter.getSplits(sft, index, sft.getUserData.get(TableSplitterOpts).asInstanceOf[String])

      case Some(p) =>
        val splitter = create(sft.getUserData.get(PartitionSplitterClass).asInstanceOf[String])
        splitter.getSplits(sft, index, p, sft.getUserData.get(PartitionSplitterOpts).asInstanceOf[String])
    }
  }

  private def create(clas: String): TableSplitter =
    if (clas == null) { DefaultSplitter.Instance } else { Class.forName(clas).newInstance().asInstanceOf[TableSplitter] }
}
